src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -129,35 +93,51 @@ private void RemoveTagsThatIntersectEdit(TextContentChangedEventArgs e)
             if (e.Changes.Count == 0)
                 return;
 
-            var buffer = e.After.TextBuffer;
-            if (!this.CachedTagTrees.TryGetValue(buffer, out var treeForBuffer))
-                return;
-
             var snapshot = e.After;
+            var buffer = snapshot.TextBuffer;
 
-            var tagsToRemove = e.Changes.SelectMany(c => treeForBuffer.GetIntersectingSpans(new SnapshotSpan(snapshot, c.NewSpan)));
-            if (!tagsToRemove.Any())
-                return;
+            // Everything we're passing in here is synchronous.  So we can assert that this must complete synchronously
+            // as well.
+            var (oldTagTrees, newTagTrees) = CompareAndSwapTagTreesAsync(
+                oldTagTrees =>
+                {
+                    if (oldTagTrees.TryGetValue(buffer, out var treeForBuffer))
+                    {
+                        var tagsToRemove = e.Changes.SelectMany(c => treeForBuffer.GetIntersectingSpans(new SnapshotSpan(snapshot, c.NewSpan)));
+                        if (tagsToRemove.Any())
+                        {
+                            var allTags = treeForBuffer.GetSpans(e.After).ToList();
+                            var newTagTree = new TagSpanIntervalTree<TTag>(
+                                buffer,
+                                treeForBuffer.SpanTrackingMode,
+                                allTags.Except(tagsToRemove, comparer: this));
```

Comment:
Unrelated to your change but I wonder if we can/should make this processing cheaper. It seems we enumerate the tree twice and allocate that intermediate list and then create a whole new tree from scratch. I assume there's a better algorithm.

(assuming this even appears in a trace)

Reply:
haven't seen it yet.  we likely can optimize more here though.